crates/cdk/src/wallet/subscription/mod.rs

```diff
@@ -0,0 +1,291 @@
+//! subscription manager that allows clients to subscribe to notifications from
+//! multiple mint servers using WebSocket or with a poll-based system, using
+//! the HTTP client.
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::sync::Arc;
+
+use tokio::sync::{mpsc, RwLock};
+use tokio::task::JoinHandle;
+use tracing::error;
+
+use crate::mint_url::MintUrl;
+use crate::nuts::nut17::Params;
+use crate::nuts::NotificationPayload;
+use crate::pub_sub::SubId;
+use crate::wallet::client::HttpClientMethods;
+
+mod http;
+#[cfg(all(
+    not(feature = "http_subscription"),
+    feature = "mint",
+    not(target_arch = "wasm32")
+))]
+mod ws;
+
+type WsSubscriptionBody = (mpsc::Sender<NotificationPayload>, Params);
+
+/// Subscription manager
+///
+/// This structure should be instantiated once per wallet at most. It is
+/// clonable since all its members are Arcs.
+///
+/// The main goal is to provide a single interface to manage multiple
+/// subscriptions to many servers to subscribe to events. If supported, the
+/// WebSocket method is used to subscribe to server-side events. Otherwise, a
+/// poll-based system is used, where a background task fetches information about
+/// the resource every few seconds and notifies subscribers of any change
+/// upstream.
+///
+/// The subscribers have a simple-to-use interface, receiving an
+/// ActiveSubscription struct, which can be used to receive updates and to
+/// unsubscribe from updates automatically on the drop.
+#[derive(Debug, Clone)]
+pub struct SubscriptionManager {
+    all_connections: Arc<RwLock<HashMap<MintUrl, SubscriptionClient>>>,
+    http_client: Arc<dyn HttpClientMethods + Send + Sync>,
+}
+
+impl SubscriptionManager {
+    /// Create a new subscription manager
+    pub fn new(http_client: Arc<dyn HttpClientMethods + Send + Sync>) -> Self {
+        Self {
+            all_connections: Arc::new(RwLock::new(HashMap::new())),
+            http_client,
+        }
+    }
+
+    /// Subscribe to updates from a mint server with a given filter
+    pub async fn subscribe(&self, mint_url: MintUrl, filter: Params) -> ActiveSubscription {
+        let subscription_clients = self.all_connections.read().await;
+        let id = filter.id.clone();
+        if let Some(subscription_client) = subscription_clients.get(&mint_url) {
+            let (on_drop_notif, receiver) = subscription_client.subscribe(filter).await;
+            ActiveSubscription::new(receiver, id, on_drop_notif)
+        } else {
+            drop(subscription_clients);
+
+            #[cfg(all(
+                not(feature = "http_subscription"),
+                feature = "mint",
+                not(target_arch = "wasm32")
+            ))]
+            let is_ws_support = self
+                .http_client
+                .get_mint_info(mint_url.clone())
+                .await
+                .map(|info| !info.nuts.nut17.supported.is_empty())
+                .unwrap_or_default();
```

Comment:
If we call this on a mint that does not include NUT17 in the info response which is possible beach it is an optional nut `is_ws_support` is set to true when we need to assume false. This leads to an infinite loop of the wallet trying to connect to the ws endpoint that does not exist on the mint. 

This happens since we have the deafult of NUT17 ` Supported settings to support all the endpoints. 
https://github.com/cashubtc/cdk/blob/7d15587e3f3b5d7e992db4b12755caf289c2e31d/crates/cdk/src/nuts/nut17/mod.rs#L38-L65

I think we should change the default here to not support any endpoints and then for out mint we won't use the Default but will explicitly set that we support them. 

It may also be worth adding a time out to the ws connection attempt so the loop does not happen. 

to reproduce this you can start a cdk-mintd from the v0.4.0 tag and then use this branch as the wallet with cdk-cli.

Reply:
It will not be false if NUT17 is not included in the info response since we use the default here https://github.com/cashubtc/cdk/blob/5ad4328a4eda7ecd7ae869db0f98799f33ab2945/crates/cdk/src/nuts/nut06.rs#L242-L245 and the default currently is that they are all supported https://github.com/cashubtc/cdk/blob/34bd97b26263b322ecfef5c92dc5d7e84c2e087c/crates/cdk/src/nuts/nut17/mod.rs#L40-L87. This is why i suggest we change the default to not supported. 

I think we should avoid the fallback in cases where we know it is not supported.